### ISSUE TITLE
$firebaseAuth.$signOut should return a promise

### ISSUE THREAD

**Issue by iRoachie** [firebase/angularfire#755]
Sign out method should return a promise similar to the normal js library method `firebase.auth().signOut()`

**Comment by katowulf** [comment]
Looks like the API document is incorrect here. signOut() is a synchronous method and returns no value. I put in a bug to fix this in the docs, but AngularFire is behaving correctly in this case.

**Comment by jwngr** [comment]
Re-opening since I do think `signOut()` in the underlying Firebase SDK is now asynchronous... AngularFire however is definitely synchronous since [we don't wait for the underlying `signOut()` call to resolve](https://github.com/firebase/angularfire/blob/4bf90a14fa588238f3d5e62ee0e439f70caa2e64/src/FirebaseAuth.js#L142), but that means we could (a) indicate that the user is signed out before they actually are and (b) eat any errors that are thrown if the user is not signed out successfully (although I'm not sure in what cases that is possible).

I'm not sure the best way to resolve this. I think we have a few options:
  1. Keep the broken, existing behavior.
  2. Change the method to return a promise and consider that a breaking change, meaning a major version bump.
  3. Change the method to return a promise and consider that a bug fix, meaning a patch version bump.

I think if we do change the method to return a promise (which would be my suggestion), we can consider that a bug fix since it won't break existing code. Existing code however is broken in the ways mentioned above. But changing the signature shouldn't make it any more or less broken.

cc/ @alfongj who may have some advice here.

**Comment by iRoachie** [comment]
I agree with this suggestion. The same behaviour should expected from the methods across the various libraries.

**Comment by jwngr** [comment]
Yeah, this (and #764 to some extent) were caused by trying to get AngularFire `2.x.x` out as soon as we could since there was a lot of community pressure to get out a version compatible with Firebase `3.x.x`. We didn't look closely enough at the new auth API and now are in a non-ideal situation. But, we will make it work. I think we need to make sure we take our time this time around to ensure we get it correct and don't cause even more API churn.

**Comment by abehaskins** [comment]
I'm not too concerned about changing `signOut()` to return a promise since if you're not expecting a return it will break nothing and if you are it will change to what you expect.

**Comment by jwngr** [comment]
Proof that `$signOut()` is in fact a promise: https://plnkr.co/edit/ZPFtJMSEpPZKXv5vwn38?p=preview

**Comment by katowulf** [comment]
Waiting on resolution to whether this is intended to be a sync or async method. Logged an internal discussion to resolve, since the Java and JS implementations differ in this detail.

**Comment by eshell** [comment]
$signInWithEmailAndPassword() is also not returning a promise for me... 
```
$firebaseAuth().$signInWithEmailAndPassword(email, password).then(function(user){
  console.log('logged in:', user);
}).catch(function(error){
  console.log(error);
});
```
I do get authenticated, but nothing happens in the then blocks. no errors or anything.

```
$firebaseAuth().$onAuthStateChanged(function(user) {
      if (user) {
        vm.user = user;
      } else {
        vm.user = null;
      }
    });
```
this works just fine to show logging in and out is successful.

**Comment by eshell** [comment]
I believe it should return a promise. $signOut().then( do something else); is pretty nice. Also, https://github.com/firebase/angularfire/issues/790

$signInWithEmailAndPassword() does not seem to return promises either.

**Comment by jwngr** [comment]
This is fixed in the just-released AngularFire [version 2.0.2](https://github.com/firebase/angularfire/releases/tag/v2.0.2). Sorry for the delay in getting it out!
